File: src/test/java/ru/jerael/booktracker/backend/application/service/EmailVerificationServiceImplTest.java
package ru.jerael.booktracker.backend.application.service;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import ru.jerael.booktracker.backend.domain.exception.NotFoundException;
import ru.jerael.booktracker.backend.domain.exception.TooManyRequestsException;
import ru.jerael.booktracker.backend.domain.mail.VerificationMailMessage;
import ru.jerael.booktracker.backend.domain.model.verification.*;
import ru.jerael.booktracker.backend.domain.repository.EmailVerificationRepository;
import ru.jerael.booktracker.backend.domain.smtp.SmtpService;
import ru.jerael.booktracker.backend.domain.validator.VerificationTokenValidator;
import ru.jerael.booktracker.backend.domain.verification.VerificationTokenGenerator;
import java.time.Duration;
import java.time.Instant;
import java.util.Optional;
import java.util.UUID;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class EmailVerificationServiceImplTest {

    @Mock
    private VerificationTokenGenerator verificationTokenGenerator;

    @Mock
    private EmailVerificationRepository emailVerificationRepository;

    @Mock
    private SmtpService smtpService;

    @Mock
    private VerificationTokenValidator verificationTokenValidator;

    @InjectMocks
    private EmailVerificationServiceImpl service;

    private final UUID userId = UUID.fromString("2c5781ea-1bc2-4561-a83d-26106df2526e");
    private final String email = "test@example.com";
    private final VerificationType type = VerificationType.REGISTRATION;
    private final String token = "123456";

    @Test
    void initiate_WhenNoExistingCode_ShouldGenerateAndSaveNewOne() {
        EmailVerificationInitiation payload = new EmailVerificationInitiation(userId, email, type);
        VerificationToken verificationToken = new VerificationToken(token, Duration.ofMinutes(10L));
        when(emailVerificationRepository.findByUserIdAndType(userId, type)).thenReturn(Optional.empty());
        when(verificationTokenGenerator.generate()).thenReturn(verificationToken);

        service.initiate(payload);

        verify(verificationTokenGenerator).generate();

        ArgumentCaptor<EmailVerification> captor = ArgumentCaptor.forClass(EmailVerification.class);
        verify(emailVerificationRepository).save(captor.capture());

        EmailVerification savedVerification = captor.getValue();
        assertEquals(email, savedVerification.email());
        assertEquals(token, savedVerification.token());
        assertNotNull(savedVerification.expiresAt());

        verify(smtpService).sendEmail(eq(email), any(VerificationMailMessage.class));
    }

    @Test
    void initiate_WhenValidCodeExists_ShouldResendSameCodeWithoutGeneratingNewOne() {
        EmailVerificationInitiation payload = new EmailVerificationInitiation(userId, email, type);
        EmailVerification found = new EmailVerification(
            UUID.randomUUID(),
            userId,
            email,
            type,
            "old token",
            Instant.now().plus(Duration.ofMinutes(8)),
            Instant.now().minus(Duration.ofMinutes(2))
        );
        when(emailVerificationRepository.findByUserIdAndType(userId, type)).thenReturn(Optional.of(found));

        service.initiate(payload);

        ArgumentCaptor<VerificationMailMessage> mailCaptor = ArgumentCaptor.forClass(VerificationMailMessage.class);
        verify(smtpService).sendEmail(eq(email), mailCaptor.capture());
        assertEquals("old token", mailCaptor.getValue().getToken().value());

        verify(verificationTokenGenerator, never()).generate();
        verify(emailVerificationRepository, never()).save(any());
    }

    @Test
    void initiate_WhenResendingTooFast_ShouldThrowException() {
        EmailVerificationInitiation payload = new EmailVerificationInitiation(userId, email, type);
        EmailVerification found = new EmailVerification(
            UUID.randomUUID(),
            userId,
            email,
            type,
            token,
            Instant.now().plus(Duration.ofMinutes(10)),
            Instant.now().minus(Duration.ofSeconds(30))
        );
        when(emailVerificationRepository.findByUserIdAndType(userId, type)).thenReturn(Optional.of(found));

        assertThrows(TooManyRequestsException.class, () -> service.initiate(payload));

        verify(smtpService, never()).sendEmail(any(), any());
        verify(emailVerificationRepository, never()).deleteByUserIdAndType(any(), any());
        verify(verificationTokenGenerator, never()).generate();
        verify(emailVerificationRepository, never()).save(any());
    }

    @Test
    void confirm_ShouldValidateAndDeleteExistingCode() {
        EmailVerificationConfirmation payload = new EmailVerificationConfirmation(userId, token, type);
        EmailVerification existing = mock(EmailVerification.class);
        when(emailVerificationRepository.findByUserIdAndType(userId, type)).thenReturn(Optional.of(existing));

        EmailVerification result = service.confirm(payload);

        assertNotNull(result);
        verify(verificationTokenValidator).validate(existing, token);
        verify(emailVerificationRepository).deleteByUserIdAndType(userId, type);
    }

    @Test
    void confirm_ShouldThrowNotFoundException_WhenRecordDoesNotExistInDb() {
        EmailVerificationConfirmation payload = new EmailVerificationConfirmation(userId, token, type);
        when(emailVerificationRepository.findByUserIdAndType(userId, type)).thenReturn(Optional.empty());

        assertThrows(NotFoundException.class, () -> service.confirm(payload));
        verify(verificationTokenValidator, never()).validate(any(), any());
    }
}

File: src/test/kotlin/ru/jerael/booktracker/backend/application/service/EmailVerificationServiceImplTest.kt
package ru.jerael.booktracker.backend.application.service

import io.mockk.*
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertNotNull
import org.junit.jupiter.api.extension.ExtendWith
import ru.jerael.booktracker.backend.domain.exception.NotFoundException
import ru.jerael.booktracker.backend.domain.exception.TooManyRequestsException
import ru.jerael.booktracker.backend.domain.mail.VerificationMailMessage
import ru.jerael.booktracker.backend.domain.model.verification.EmailVerification
import ru.jerael.booktracker.backend.domain.repository.EmailVerificationRepository
import ru.jerael.booktracker.backend.domain.smtp.SmtpService
import ru.jerael.booktracker.backend.domain.validator.VerificationTokenValidator
import ru.jerael.booktracker.backend.domain.verification.VerificationTokenGenerator
import ru.jerael.booktracker.backend.factory.verification.EmailVerificationDomainFactory
import java.time.Instant
import java.util.*

@ExtendWith(MockKExtension::class)
class EmailVerificationServiceImplTest {
    
    @MockK
    private lateinit var verificationTokenGenerator: VerificationTokenGenerator
    
    @MockK
    private lateinit var emailVerificationRepository: EmailVerificationRepository
    
    @MockK(relaxed = true)
    private lateinit var smtpService: SmtpService
    
    @MockK
    private lateinit var verificationTokenValidator: VerificationTokenValidator
    
    @InjectMockKs
    private lateinit var service: EmailVerificationServiceImpl
    
    @Test
    fun `when code not existing, initiate should generate and save new one`() {
        val payload = EmailVerificationDomainFactory.createEmailVerificationInitiation()
        val verificationToken = EmailVerificationDomainFactory.createVerificationToken()
        
        every { emailVerificationRepository.findByUserIdAndType(payload.userId, payload.type) } returns Optional.empty()
        every { verificationTokenGenerator.generate() } returns verificationToken
        every { emailVerificationRepository.save(any()) } returns mockk()
        
        service.initiate(payload)
        
        verify { verificationTokenGenerator.generate() }
        
        val emailVerificationSlot = slot<EmailVerification>()
        verify { emailVerificationRepository.save(capture(emailVerificationSlot)) }
        
        with(emailVerificationSlot.captured) {
            assertEquals(payload.email, email)
            assertEquals(verificationToken.value, token)
            assertNotNull(expiresAt)
        }
        
        verify {
            smtpService.sendEmail(eq(payload.email), any(VerificationMailMessage::class))
        }
    }
    
    @Test
    fun `when valid code exists, initiate should resend same code without generating new one`() {
        val payload = EmailVerificationDomainFactory.createEmailVerificationInitiation()
        
        val userId = payload.userId
        val email = payload.email
        val type = payload.type
        
        val oldToken = "old token"
        
        val found = EmailVerificationDomainFactory.createEmailVerification(
            userId = userId,
            email = email,
            type = type,
            token = oldToken,
            createdAt = Instant.now().minusSeconds(120)
        )
        
        every { emailVerificationRepository.findByUserIdAndType(userId, type) } returns Optional.of(found)
        
        service.initiate(payload)
        
        val verificationMailMessageSlot = slot<VerificationMailMessage>()
        verify { smtpService.sendEmail(eq(email), capture(verificationMailMessageSlot)) }
        
        assertEquals(oldToken, verificationMailMessageSlot.captured.token.value)
        
        verify(exactly = 0) { verificationTokenGenerator.generate() }
        verify(exactly = 0) { emailVerificationRepository.save(any()) }
    }
    
    @Test
    fun `when resending too fast, initiate should throw TooManyRequestsException`() {
        val payload = EmailVerificationDomainFactory.createEmailVerificationInitiation()
        
        val userId = payload.userId
        val email = payload.email
        val type = payload.type
        
        val found = EmailVerificationDomainFactory.createEmailVerification(
            userId = userId,
            email = email,
            type = type,
            createdAt = Instant.now().minusSeconds(30)
        )
        
        every { emailVerificationRepository.findByUserIdAndType(userId, type) } returns Optional.of(found)
        
        assertThrows(TooManyRequestsException::class.java) { service.initiate(payload) }
        
        verify(exactly = 0) { smtpService.sendEmail(any(), any()) }
        verify(exactly = 0) { emailVerificationRepository.deleteByUserIdAndType(any(), any()) }
        verify(exactly = 0) { verificationTokenGenerator.generate() }
        verify(exactly = 0) { emailVerificationRepository.save(any()) }
    }
    
    @Test
    fun `confirm should validate and delete existing code`() {
        val payload = EmailVerificationDomainFactory.createEmailVerificationConfirmation()
        val existing = mockk<EmailVerification>()
        
        val userId = payload.userId
        val type = payload.type
        
        every { emailVerificationRepository.findByUserIdAndType(userId, type) } returns Optional.of(existing)
        every { verificationTokenValidator.validate(any(), any()) } just Runs
        every { emailVerificationRepository.deleteByUserIdAndType(any(), any()) } just Runs
        
        val result = service.confirm(payload)
        
        assertNotNull(result)
        
        verify { verificationTokenValidator.validate(existing, payload.token) }
        verify { emailVerificationRepository.deleteByUserIdAndType(userId, type) }
    }
    
    @Test
    fun `when record does not exist in database, confirm should throw NotFoundException`() {
        val payload = EmailVerificationDomainFactory.createEmailVerificationConfirmation()
        
        every { emailVerificationRepository.findByUserIdAndType(any(), any()) } returns Optional.empty()
        
        assertThrows(NotFoundException::class.java) { service.confirm(payload) }
        
        verify(exactly = 0) { verificationTokenValidator.validate(any(), any()) }
    }
}